# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/mediasink/VideoSink.cpp
# Commit: 35d86243cb45
# Full Hash: 35d86243cb45f7e02dd6ec0eea04cd62c2a70196
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2024-09-16 09:14:26
# Regressor Bug: 1908245
# File Overlap Count: 1
# Description:
#   Bug 1908245 - Add support for missing requestVideoFrameCallback parameters. r=media-playback-reviewers,webidl,padenot,saschanaz
#   
#   This patch adds support for exposing the capture time, receive time,
#   processing duration and RTP timestamp parameters on the callback for
#   rVFC for WebRTC related elements.
# ==============================================================================

diff -r 7d143d92bf2a -r 35d86243cb45 dom/media/mediasink/VideoSink.cpp
--- a/dom/media/mediasink/VideoSink.cpp	Sun Sep 15 20:03:40 2024 +0000
+++ b/dom/media/mediasink/VideoSink.cpp	Sun Sep 15 21:49:38 2024 +0000
@@ -349,9 +349,12 @@
       video->mImage = mBlankImage;
     }
     video->MarkSentToCompositor();
-    mContainer->SetCurrentFrame(video->mDisplay, video->mImage, now);
+    mContainer->SetCurrentFrame(video->mDisplay, video->mImage, now,
+                                media::TimeUnit::Invalid(), video->mTime);
     if (mSecondaryContainer) {
-      mSecondaryContainer->SetCurrentFrame(video->mDisplay, video->mImage, now);
+      mSecondaryContainer->SetCurrentFrame(video->mDisplay, video->mImage, now,
+                                           media::TimeUnit::Invalid(),
+                                           video->mTime);
     }
     return;
   }
@@ -362,10 +365,14 @@
 
   RefPtr<Image> blank =
       mContainer->GetImageContainer()->CreatePlanarYCbCrImage();
-  mContainer->SetCurrentFrame(aInfo.mDisplay, blank, now);
+  mContainer->SetCurrentFrame(aInfo.mDisplay, blank, now,
+                              media::TimeUnit::Invalid(),
+                              media::TimeUnit::Invalid());
 
   if (mSecondaryContainer) {
-    mSecondaryContainer->SetCurrentFrame(aInfo.mDisplay, blank, now);
+    mSecondaryContainer->SetCurrentFrame(aInfo.mDisplay, blank, now,
+                                         media::TimeUnit::Invalid(),
+                                         media::TimeUnit::Invalid());
   }
 }
 
@@ -476,6 +483,7 @@
     }
     img->mFrameID = frame->mFrameID;
     img->mProducerID = mProducerID;
+    img->mMediaTime = frame->mTime;
 
     VSINK_LOG_V("playing video frame %" PRId64
                 " (id=%x, vq-queued=%zu, clock=%" PRId64 ")",
@@ -672,11 +680,14 @@
     // that in the secondary container as well.
     AutoLockImage lockImage(mainImageContainer);
     TimeStamp now = TimeStamp::Now();
-    if (RefPtr<Image> image = lockImage.GetImage(now)) {
+    if (const auto* owningImage = lockImage.GetOwningImage(now)) {
       AutoTArray<ImageContainer::NonOwningImage, 1> currentFrame;
       currentFrame.AppendElement(ImageContainer::NonOwningImage(
-          image, now, /* frameID */ 1,
-          /* producerId */ ImageContainer::AllocateProducerID()));
+          owningImage->mImage, now, /* frameID */ 1,
+          /* producerId */ ImageContainer::AllocateProducerID(),
+          owningImage->mProcessingDuration, owningImage->mMediaTime,
+          owningImage->mWebrtcCaptureTime, owningImage->mWebrtcReceiveTime,
+          owningImage->mRtpTimestamp));
       secondaryImageContainer->SetCurrentImages(currentFrame);
     }
   }